MobiusCore/Source/EffectHandlers/EffectRouter.swift

```diff
@@ -24,33 +24,46 @@ public struct EffectRouter<Input, Output> {
         routes = []
     }
 
-    private init(routes: [Route<Input, Output>]) {
+    fileprivate init(routes: [Route<Input, Output>]) {
         self.routes = routes
     }
 
-    public func add<Payload>(
-        path: EffectPath<Input, Payload>,
-        to handler: EffectHandler<Payload, Output>
-    ) -> EffectRouter<Input, Output> {
-        let route = Route(path: path, handler: handler)
-        return EffectRouter(routes: self.routes + [route])
+    /// Add a route for effects which satisfy `payload`. `payload` is a function which returns an optional value for a given effect. If this value is non-`nil`
+    /// this route will be taken with that non-`nil` value as input. A different route will be taken if `nil` is returned.
+    /// - Parameter payload: a function which returns a non-`nil` value if this route should be taken, and `nil` if a different route should be taken.
+    public func route<Payload>(
+        payload: @escaping (Input) -> Payload?
+    ) -> PartialEffectRouter<Input, Payload, Output> {
+        return PartialEffectRouter(routes: routes, path: payload)
     }
 
     public var asConnectable: AnyConnectable<Input, Output> {
         return compose(routes: routes)
     }
 }
 
+public struct PartialEffectRouter<Input, Payload, Output> {
```

Comment:
The "primitive" operation on a router is to add a payload extracting route to an effect handler. All of the DSL functions are building on those two primitive operations.
I think we can explore moving these around at a later stage. It would be a compatible change from a user's perspective.